src/Raven.Server/Documents/Sharding/ShardedDocumentsStorage.cs

```diff
@@ -128,31 +159,122 @@ public IEnumerable<ReplicationBatchItem> GetTombstonesByBucketFrom(DocumentsOper
+                break;
+
+            var docCv = context.GetChangeVector(document.ChangeVector);
+            if (ChangeVectorUtils.GetConflictStatus(docCv, upTo) != ConflictStatus.AlreadyMerged)
+                break;
+
+            // check change vectors of all document extensions
+            if (HasDocumentExtensionWithGreaterChangeVector(context, document.LowerId, upTo)) 
+                break;
+            
+            Delete(context, document.Id, document.Flags | DocumentFlags.Artificial);
+
+            // delete revisions for document
+            RevisionsStorage.DeleteRevisionsFor(context, document.Id, artificial: true);
+
+            numOfDeletions++;
         }
+
+        return numOfDeletions;
     }
 
-    public static long DeleteItemsByBucketForDocuments(DocumentsOperationContext context, CollectionName collection, int bucket, ChangeVector upTo)
+    private void MarkTombstonesAsArtificial(DocumentsOperationContext context, int bucket)
     {
-        var table = context.Transaction.InnerTransaction.OpenTable(DocsSchema, collection.GetTableName(CollectionTableType.Documents));
-        return DeleteItemsByBucket(context, table, DocsSchema.DynamicKeyIndexes[CollectionDocsBucketAndEtagSlice], (int)DocumentsTable.ChangeVector, bucket, upTo);
+        var readTable = new Table(TombstonesSchema, context.Transaction.InnerTransaction);
+
+        foreach (var result in GetItemsByBucket(context.Allocator, readTable, TombstonesSchema.DynamicKeyIndexes[TombstonesBucketAndEtagSlice], bucket, etag: 0))
+        {
+            var tombstone = TableValueToTombstone(context, ref result.Result.Reader);
+            if (tombstone.Flags.Contain(DocumentFlags.Artificial))
+                continue;
+
+            var collection = TableValueToId(context, (int)TombstoneTable.Collection, ref result.Result.Reader);
+            if (_collectionsCache.TryGetValue(collection, out var collectionName) == false)
+            {
+                collectionName = new CollectionName(collection);
+                _collectionsCache.TryAdd(collection, collectionName);
+            }
+            
+            var writeTable = context.Transaction.InnerTransaction.OpenTable(TombstonesSchema, collectionName.GetTableName(CollectionTableType.Tombstones));
+
+            var newEtag = GenerateNextEtag();
+            var cv = ChangeVector.Merge(context.LastDatabaseChangeVector, context.GetChangeVector(tombstone.ChangeVector), context);
+            var flags = tombstone.Flags | DocumentFlags.Artificial;
+
+            using (Slice.From(context.Allocator, cv, out var cvSlice))
+            using (Slice.External(context.Allocator, tombstone.LowerId, out var keySlice))
+            using (DocumentIdWorker.GetStringPreserveCase(context, collection, out Slice collectionSlice))
+            using (writeTable.Allocate(out TableValueBuilder tvb))
+            {
+                var clonedKey = keySlice.Clone(context.Allocator);
```

Comment:
can we deallocate after writeTable.Update? Leak?

Reply:
done